driver/src/metrics/metrics_server.rs

```diff
@@ -0,0 +1,42 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use hyper::service::{make_service_fn, service_fn};
+use hyper::{Body, Error, Response, Server};
+use prometheus::{Encoder, Registry, TextEncoder};
+
+pub struct MetricsServer {
+    registry: Arc<Registry>,
+}
+
+impl MetricsServer {
+    pub fn new(registry: Arc<Registry>) -> Self {
+        Self { registry }
+    }
+    pub fn serve(&self, port: u16) -> Pin<Box<dyn Future<Output = Result<(), Error>> + Send>> {
+        let addr = ([0, 0, 0, 0], port).into();
+        let registry = self.registry.clone();
+        let service = make_service_fn(move |_| {
+            let registry = registry.clone();
```

Comment:
Is there any way to get around all these registry clones?

Reply:
I have tried this a bit. Afaict we would need to implement hyper's `Service` trait manually but its not worth the complexity for something where a small change in performance does not matter.